src/text/encoder.rs

```diff
@@ -0,0 +1,264 @@
+            Timestamp => "null.timestamp",
+            Symbol => "null.symbol",
+            String => "null.string",
+            Blob => "null.blob",
+            Clob => "null.clob",
+            List => "null.list",
+            SExpression => "null.sexp",
+            Struct => "null.struct",
+        }
+    }
+
+    #[inline(always)]
+    fn bool_text(value: bool) -> &'static str {
+        match value {
+            true => "true",
+            false => "false",
+        }
+    }
+
+    /// Writes a null of the specified Ion type to `output`.
+    pub fn encode_null<W: io::Write>(output: &mut W, ion_type: IonType) -> IonResult<()> {
+        write!(output, "{}", TextEncoder::null_text(ion_type))?;
+        Ok(())
+    }
+
+    /// Writes a bool value to `output` as an Ion boolean.
+    pub fn encode_bool<W: io::Write>(output: &mut W, value: bool) -> IonResult<()> {
+        write!(output, "{}", TextEncoder::bool_text(value))?;
+        Ok(())
+    }
+
+    /// Writes an i64 value to `output` as an Ion integer.
+    pub fn encode_i64<W: io::Write>(output: &mut W, value: i64) -> IonResult<()> {
+        write!(output, "{}", value)?;
+        Ok(())
+    }
+
+    /// Writes a BigDecimal value to `output` as an Ion decimal.
+    pub fn encode_big_decimal<W: io::Write>(output: &mut W, value: &BigDecimal) -> IonResult<()> {
+        write!(output, "{}", &value)?;
+        Ok(())
+    }
+
+    /// Writes an f64 value to `output` as an Ion float.
+    pub fn encode_f64<W: io::Write>(output: &mut W, value: f64) -> IonResult<()> {
+        if value.is_nan() {
+            write!(output, "nan")?;
+            return Ok(());
+        }
+
+        if value.is_infinite() {
+            if value.is_sign_positive() {
+                write!(output, "+inf")?;
+            } else {
+                write!(output, "-inf")?;
+            }
+            return Ok(());
+        }
+
+        // The {:e} formatter provided by the Display trait writes floats using scientific
+        // notation. It works for all floating point values except -0.0 (it drops the sign).
+        // See: https://github.com/rust-lang/rust/issues/20596
+        if value == 0.0f64 && value.is_sign_negative() {
+            write!(output, "-0e0")?;
+            return Ok(());
+        }
+
+        write!(output, "{:e}", value)?;
+        Ok(())
+    }
+
+    /// Writes a DateTime value to `output` as an Ion timestamp.
+    pub fn encode_datetime<W: io::Write, I: Into<DateTime<FixedOffset>>>(
+        output: &mut W,
+        value: I,
+    ) -> IonResult<()> {
+        let fixed_offset_datetime: DateTime<FixedOffset> = value.into();
+        write!(output, "{}", fixed_offset_datetime.to_rfc3339())?;
+        Ok(())
+    }
+
+    /// Writes a string value to `output` as an Ion symbol.
+    pub fn encode_str_as_symbol<S: AsRef<str>, W: io::Write>(
+        output: &mut W,
+        value: S,
+    ) -> IonResult<()> {
+        write!(output, "'{}'", value.as_ref())?;
+        Ok(())
+    }
```

Comment:
Note that this API has an interesting implication of consuming `value` (might not be what the user wants if they pass a `String`), now looking down at the writer, you already only take in `&str` so it is a trivial copy, but then I have to ask, why are we being generic here and not just make it `&str`?

To illustrate my point:

```rust
use std::convert::*;

fn print_str<S: AsRef<str>>(s: S) {
    println!("{}", s.as_ref());
}

fn main() {
    let s = String::from("blah");
    print_str(s);
    
    // Cannot do this! s is moved into `print_str`.
    //print_str(s);
}
```

Reply:
I am okay with `AsRef<str>` here after talking to you offline--I think that it is "free" so that works well.